### ISSUE TITLE
clarify naming rules

### ISSUE THREAD

**Issue by ahwagner** [ga4gh/vrsatile#25]
> forget this. we can address later if necessary. I'm getting a bit confused by our naming approach. Should the class name include the type of VRS object it measures (Variation, Location, Sequence, Gene, Feature, etc...) or should it contain the core elements that the rule uses to do the measuring to identify the VRS objects that it will output. Or both?  Will we have LocationRules and VariationRules, etc..? If so, we should try to define a rational approach to how we name these "categorical" classes.

_Originally posted by @larrybabb in https://github.com/ga4gh/vrsatile/pull/22#discussion_r753306924_

**Comment by ahwagner** [comment]
I agree this should be formalized. I propose that we name the rule based on the concept it represents instead of by the constituent features, similar to how we do it in VRS (e.g. `Allele` instead of `SequenceExpressionLocation` / `LocationSequenceExpression`).

Better to do it here in a GitHub thread, and then implement in code after agreement.

I think this should start with a top-level class, which I think should be `CategoricalVariationDomain`:
> A representation of a categorically-defined variation domain, in which individual variation instances may be members.

Under this there are several concrete subclasses. The first two of these are `Canonical` and `Complex`.

`Canonical`:
> A categorical variation domain characterized by a canonical Variation context to which members lift-over, project, translate, or otherwise directly align.

`Complex`: 
> A categorical variation domain jointly characterized by two or more other categorical variation domains.

**Comment by larrybabb** [comment]
I like the definitions for the 3 classes above. I think the terms `Canonical` and `Complex` by themselves are not adequate to name a class but if they are modifiers for `Variation` or `Location` then I think these terms are fine. For example, `CanonicalVariation` is a kind of `CategoricalVariation` as well as `ComplexVariation`.  

I was wondering if we would ever go down the path of `CategoricalLocation` classes? In some ways, maybe in every way, `Features` are `CanonicalLocations` (I think).  When humans and computer systems refer to `Genes` they typically consider any contextual reference to a gene to be representative of that gene. I believe the same may be true of all `Features` (`NamedLocations` or `NamedRegions`).
